### ISSUE TITLE
On Windows, follow Windows patterns/practices.

### ISSUE THREAD

**Issue by MicahZoltu** [OpenVPN/openvpn-gui#179]
Configuration data for installed applications on Windows should _NOT_ go in `%USERPROFILE%`.  It should instead go into `%APPDATA%/<vendor>`.  At the moment on Windows, OpenVPN installs its configuration into `%USERPROFILE%/OpenVPN/`.  This folder is meant to only store files intentionally saved here by the user (e.g., via a save dialog box).  Applications should _not_ be writing arbitrary files to this location.

Please follow Windows best practices and move the config location to `%APPDATA%/OpenVPN/`.

**Comment by ffes** [comment]
I fully agree with you.

Small problem is though that the current implementation is already used in the wild (at our office, for instance). So the current location `%USERPROFILE%/OpenVPN/config/*.ovpn` should be searched as well (at least for one 2.x release), so current deployments don't break.

**Comment by mattock** [comment]
We actually discussed the config file directory's location quite a bit. We would have preferred ```%APPDATA%\OpenVPN``` but decided not to use it as Windows hides ```%APPDATA%``` from directory listings (correct?). So we thought that following best practices in this case would have increased our support burden considerably. If I recall correctly the problem can be fixed with filesystem ACLs, but we did not want to mess with those, either.

**Comment by MicahZoltu** [comment]
Often the way applications deal with the invisibleness of `AppData` is to put a shortcut to their configuration directory in the start menu.  I think a more elegant solution is to have a menu item in the application for "open configuration directory in explorer".  Really though, if the average user needs to be opening the configuration directory that strikes me as a bigger problem and suggests that the UI is lacking.  I believe there is already an outstanding issue for making it so users can easily add configuration files via the UI rather than file explorer, so perhaps once that is addressed the expectation that the user will need to open a config folder for anything but extreme cases will go away?

**Comment by mattock** [comment]
@MicahZoltu : the workarounds you suggest sound reasonable. Once the pull request [Functionality to import configuration file from the command line](https://github.com/OpenVPN/openvpn-gui/pull/76) has been merged, the location of config files will matter less than now. That said, importing OpenVPN config files is pretty tricky when there are references to external files (e.g. certificates, scripts, auth-user-pass files). We will probably always need an option to import config files manually, and for that proper documentation and the workarounds you suggest are probably good enough.

**Comment by selvanair** [comment]
I think until the GUI has a proper config editor that can handle dependent files, create configs from a dialog, edit scripts etc., the users will have to directly manipulate files in the config folder. Hence the need to treat them like personal documents kept in a "visible" location.

**Comment by chipitsine** [comment]
creating configs from scratch is a bit weird.

for example, iphone, android do not have anything, just "import" from url
and everybody is hust happy with that.

**Comment by jupsoleil** [comment]
I think it adds some extra security of the ovpn file is stored in %USERPROFILE%\OpenVPN\Config since this ovpn filed is linked to the user, not to the machine.  Or at least it is such in my domain.  So I prefer that other users who log in don't even see that there's an ovpn available if they didn't get a personal one.
People who prefer ovpns linked to a machine can still use %PROGRAMFILES%\OpenVPN\Config.

**Comment by MicahZoltu** [comment]
@jupsoleil `%AppData%` is per user.  Also, machine wide config should go into `ProgramData%`, not `%ProgramFiles%`

**Comment by jupsoleil** [comment]
@MicahZoltu indeed you are right, I overlooked that!  So your proposal is sound 👍 .

**Comment by Kunzol** [comment]
I found a document describing it quiet clearly.
[Developers best practice for Windows user profiles](https://www.hass.de/content/developers-best-practice-windows-user-profiles)

%AppData% would be best, but be aware, that this could be an UNC path in a roaming profile environment. Which is most probably a network drive.
Software like OpenVPN which modifies the network connection of the system could break the access to the directory.

**Comment by SergeyVSB** [comment]
Well, there is %LOCALAPPDATA%

**Comment by MicahZoltu** [comment]
@selvanair Can you shed some light on what discussions resulted in this getting closed as not planned?  The conversation above seems to imply that there was weak support for fixing this once the UI had better import options.

The current behavior is still wrong and I think there *should* be a plan to fix it eventually.  I'm generally OK with projects taking the time they need to get things fixed (in this case, more than 5 years it seems) but I don't think this should just be left unfinished forever.

**Comment by selvanair** [comment]
The premise of this thread that we are not following best practices by keeping configs in user-visible folders stems from a misunderstanding that config files are internal assets of this application. That's not the case:

(1) In the current design of the GUI, config files are user's assets --- something they can create, delete and edit outside of the application using independent tools --- not some internal configuration that belongs to the application.

(2) We do not have a fully functional config import feature that will automatically install not just config files but their dependencies too -- cert, key files, scripts etc. Users are required to copy those files into the config folder manually. So, "hiding" user configs in a hard to find location like %....APPDATA% is not appropriate for this application in its current state.

If interested in addressing (2) open a separate issue or PR.

**Comment by MicahZoltu** [comment]
In that case, if these are truly user assets, then a save/open dialog should be used to let the user choose where to read/write files to.  The idea is that `%USERPROFILE%` is a directory the user controls and they can *choose* to write there or write directly, in a subfolder, or elsewhere.  Apps should not be hard coding any leads/writes to that directory.

If you want a fixed location that the app decides/controls structure of, then `APPDATA/<product>` or `LOCALAPPDATA/<product>`.

If you want user controlled files, then an open/save dialog which can reasonably default to `USERPROFILE` and you can remember their previous location for future dialog defaults or auto-reply (e.g., recent items list, reload on restart).

**Comment by cron2** [comment]
Hi,

On Sat, Dec 17, 2022 at 08:53:22AM -0800, Micah Zoltu wrote:
> If you want user controlled files, then an open/save dialog which can reasonably default to `USERPROFILE` and you can remember their previous location for future dialog defaults or auto-reply (e.g., recent items list, reload on restart).

We don't have resources to work on this any time soon.

If someone else wants to take a stab at contributing code, that might
get things moving faster.

gert

-- 
"If was one thing all people took for granted, was conviction that if you 
 feed honest figures into a computer, honest figures come out. Never doubted 
 it myself till I met a computer with a sense of humor."
                             Robert A. Heinlein, The Moon is a Harsh Mistress

Gert Doering - Munich, Germany                             ***@***.***

**Comment by JackTrapper** [comment]
Currently the OpenVPN application settings end up in:

- 📁`%USERPROFILE%`\OpenVPN
  - 📁config
    - 📁My_Work_VPN
       -  📄My_Work_VPN.ovpn
  - 📁log
     -  📄My_Work_VPN.log

On Windows (since 1995) this has not been the appropriate folder for these files. This is *application data*:

- `FOLDERID_LocalAppData`
- `CSIDL_LOCAL_APPDATA`
- `%LOCALAPPDATA%`
- `%USERPROFILE%`\AppData\Local *(for example)*

And it should be placed in:

- 📁`%LOCALAPPDATA%`\OpenVPN
  - 📁config
    - 📁My_Work_VPN
       -  📄My_Work_VPN.ovpn
  - 📁log
     -  📄My_Work_VPN.log
    

If you feel these files should roam with the user, as they logon to different workstations in the enterprise, that their VPN settings should roam with them, you can put these files in their *"roaming"* **AppData** folder (rather than the **LocalAppData** folder):

- `FOLDERID_RoamingAppData`
- `CSIDL_APPDATA`
- `%APPDATA%`
- `%USERPROFILE%`\AppData\Roaming *(for example)*

- 📁`%LOCALAPPDATA%`\OpenVPN
  - 📁config
    - 📁My_Work_VPN
       -  📄My_Work_VPN.ovpn
  - 📁log
     -  📄My_Work_VPN.log


Placing these files in the user's profile folder (*i.e.* their home directory) is not appropriate. Windows has *defined* the place where **APPDATA** should go. 

- It helps with backup/restore (knowing that all your application data is collected in one place)
- for moving the location of the **AppData** folder to another path *(perhaps an off the space constrained M.2 to a spinning platter)*
- and for being a good tenet on the user's personal computer

From the **Designed for Microsoft Windows Application Specification**, there is a section specifically dedicated to how to decide where to place files:

3.2	Classify and store application data correctly
================================

Applications typically work with two fundamental types of documents: those that are created by the user, and those that are created for use by the application itself. For user-created data, please refer to the guidelines in section 3.1. Section 3.2 applies directly to the data that is created by an application to store user-specific information. This data is used by the application only, and is never intended to be accessed directly by the user.

By storing this application-specific data in one of the several valid locations as described by this specification, you make it possible for multiple people to use the same computer without corrupting or improperly modifying each other’s data. The specification provides several valid locations and you are free to choose the location that works best for your needs.

A clear benefit to the developer is that compliance with this requirement can actually result in fewer lines of code. There is a single API call, **SHGetFolderPath**, that enables you to determine the correct location in which to store the user’s data and the user-specific application data.

Classifying and storing application data according to the guidelines in this requirement provides these benefits:

•	It enables multiple family members to share a computer and helps enable Fast User Switching. 
•	It enables business-related operations such as roaming, off-line storage, and allowing the operating system and its applications to be secured. 
•	It ensures a consistent and abstracted location for user data, enforces per-user separation of application data.
•	It is one of the key factors in enabling remote use of the application.

Application data includes user preferences, application state, temp files, and so on. 

> **IMPORTANT**: For files that the user can open (such as documents, music, pictures, and so on), use the guidelines in section 3.1.

This section identifies the valid file folders and the valid registry locations that applications must use for this data, and gives guidance on how to choose which of these locations are best used in different circumstances. The choice of valid locations to use is left to the software developer.

Classify application data into the following categories:

•	Per user, roaming
•	Per user, non-roaming
•	Per computer (non-user specific and non-roaming)

> **NOTE**:  There may be more than one category for the different application data stored by your application.

For applications not intended to be used in a domain environment (most games and home products for example), classifying the application data as per user, non-roaming might be an appropriate choice.

It is best to use application data file folders rather than the registry for storing application data in excess of 64K. The registry is an acceptable choice for small amounts of data. At installation time, try to store less than a total of 128K across `HKEY_CURRENT_USER` (`HKCU`) and `HKEY_LOCAL_MACHINE` (`HKLM`). 

To comply with this specification, store application data files appropriately as either common or per-user. That is:

•	In a subfolder of either the common application folder (identified by `CSIDL_COMMON_APPDATA`), or 
•	In the user profile folders: application data (`CSIDL_APPDATA`) or local application data (`CSIDL_LOCAL_APPDATA`). 

The subfolder to create to store user data files in is:

> [company name]\[product name]\[version].


tl;dr: Only users should be placing folders in **their** profile folder. *Their* home directory. Not yours.
